Add set_data_3d and get_data_3d to Line3d

Raab70 opened the pull request:
## PR Summary
Fixes #8914 by adding set_data_3d and get_data_3d methods to Line3D. This allows for setting and getting of 3d data without overriding the widely used set_data and get_data methods. 

```
> from mpl_toolkits.mplot3d import Axes3D
> x = [0, 1]
> y = [0, 1]
> z = [0, 1]
> fig = plt.figure()
> ax = fig.add_subplot(111, projection='3d')
> line = ax.plot(x,y,z)[0]
> line.get_data_3d()
(array([0, 1]), array([0, 1]), array([0, 1]))
> line.set_data_3d([0,2], [0,2], [0,2])
> line.get_data_3d()
([0, 2], [0, 2], [0, 2])
```


I looked for testing and documentation and did not see any testing or documentation related to Line3D would require updating. Please let me know if this is incorrect. 

## PR Checklist

- [x] Has Pytest style unit tests
- [X] Code is PEP 8 compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

tacaswell commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -130,6 +130,27 @@ def set_3d_properties(self, zs=0, zdir='z'):
+        """
+        Return the xdata, ydata, zdata.
+        """
+        return self._verts3d
```
Does this need to un-juggle the order of the data?

Raab70 replied:
It probably should. We're calling `set_3d_properties` with no `zdir` so we shouldn't do any juggling but the user could change the zdir, I'll probably also allow an optional kwarg to allow the user to also set zdir right from set_data_3d.

tacaswell replied:
Both should probably get the `zdir` kwarg?

WeatherGod replied:
The juggle axis feature is for converting a 2D object into 3D. It essentially states which dimension is going to be a constant. It is not applicable here.

WeatherGod commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -130,6 +130,27 @@ def set_3d_properties(self, zs=0, zdir='z'):
+            x, y, z = args
+
+        self.set_xdata(x)
+        self.set_ydata(y)
```
This is incorrect. `set_xdata()` and `set_ydata()` are for the 2D projected values. `self._verts3d` is just a tuple of the 3 1D arrays.

tacaswell replied:
Ah, fair enough.  Using this method is very suggestive due to

https://github.com/matplotlib/matplotlib/blob/37978a8a701530acc824500c7d4a09cad1d93d7c/lib/mpl_toolkits/mplot3d/art3d.py#L117-L129

but on a bit closer inspection it seems that these methods are used in the 2D class -> 3D class upcasting, not as a standard part of the setup.

Raab70 commented:
I think I've addressed all issues here but if anyone wants to let me know why Travis is complaining I'd love to hear. From the log it looks like it failed in `lib/matplotlib/tests/test_compare_images.py:222` which is a file not changed in this PR.

WeatherGod commented:
Try rebasing on master and force-pushing. There have been some fixes to
Travis testing the past few weeks.

On Sun, Feb 11, 2018 at 11:57 PM, Rob Harrigan <notifications@github.com>
wrote:

> I think I've addressed all issues here but if anyone wants to let me know
> why Travis is complaining I'd love to hear. From the log it looks like it
> failed in lib/matplotlib/tests/test_compare_images.py:222 which is a file
> not changed in this PR.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/10111#issuecomment-364829251>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-GVB65SiP8Fdsxh3xzQu-5sepSU1ks5tT8S_gaJpZM4RNLt3>
> .
>

Raab70 commented:
Thanks for the tip @WeatherGod, that seems to have fixed Travis but broken circleci...

Again, definitely not something changed in this PR. 

```
Warning, treated as error:
/home/circleci/project/lib/matplotlib/image.py:docstring of matplotlib.image.NonUniformImage.set_interpolation:10:Inline substitution_reference start-string without end-string.
```

Raab70 commented:
Ping @tacaswell and @WeatherGod, rebasing on master seems to have fixed it this time.

themightyoarfish commented:
Can this be merged?

WeatherGod commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -151,6 +151,39 @@ def set_3d_properties(self, zs=0, zdir='z'):
+        if len(args) == 1:
+            self._verts3d = args[0]
+        else:
+            self._verts3d = args
```
I think you need to add `self.stale = True` here.
